community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/DuplicateCheckStrategy.java

```diff
@@ -0,0 +1,274 @@
+    {
+        private Map<Object,Long> valueNodeIdMap;
+
+        MapDuplicateCheckStrategy( int expectedNumberOfEntries )
+        {
+            this.valueNodeIdMap = new HashMap<>( expectedNumberOfEntries );
+        }
+
+        @Override
+        public void checkForDuplicate( Property[] properties, Object[] values, long nodeId )
+                throws IndexEntryConflictException
+        {
+            Long previousNodeId = valueNodeIdMap.put( new PropertyValues( properties, values ), nodeId );
+            if ( previousNodeId != null )
+            {
+                throw new IndexEntryConflictException( previousNodeId, nodeId,
+                        OrderedPropertyValues.ofUndefined( values ) );
+            }
+        }
+
+        @Override
+        void checkForDuplicate( Property property, Object value, long nodeId ) throws IndexEntryConflictException
+        {
+            Long previousNodeId = valueNodeIdMap.put( property, nodeId );
+            if ( previousNodeId != null )
+            {
+                throw new IndexEntryConflictException( previousNodeId, nodeId, value );
+            }
+        }
+
+        private static class PropertyValues
+        {
+            private final Property[] properties;
+            private final Object[] values;
+
+            PropertyValues( Property[] properties, Object[] values )
+            {
+                this.properties = properties;
+                this.values = values;
+            }
+
+            @Override
+            public boolean equals( Object o )
+            {
+                if ( this == o )
+                {
+                    return true;
+                }
+                if ( o == null || getClass() != o.getClass() )
+                {
+                    return false;
+                }
+
+                PropertyValues that = (PropertyValues) o;
+                return propertyValuesEqual( properties, that.values );
+            }
+
+            @Override
+            public int hashCode()
+            {
+                int result = 0;
+                for ( Property property : properties )
+                {
+                    result = 31 * (result + property.hashCode());
+                }
+                return result;
+            }
+        }
+    }
+
+    /**
+     * Strategy that uses arrays to store entries and uses hash codes to split those entries over different buckets.
+     * Number of buckets and size of entries block are dynamic and evaluated based on expected number of duplicates.
+     */
+    static class BucketsDuplicateCheckStrategy extends DuplicateCheckStrategy
+    {
+        private static final int BASE_ENTRY_SIZE = 1000;
+        private static final int DEFAULT_BUCKETS = 10;
+        static final int BUCKET_STRATEGY_ENTRIES_THRESHOLD = BASE_ENTRY_SIZE * DEFAULT_BUCKETS;
+
+        private static final int MAX_NUMBER_OF_BUCKETS = 100;
+        private final int numberOfBuckets;
+        private EntrySet[] actualValues;
```

Comment:
I was a bit confused about what was the "buckets". Can we improve naming to make this more clear?

Something like (but probably not):
`actualValues` -> `buckets`
`EntrySet` -> `BucketNode`
`entrySetSize` -> `bucketNodeSize`